x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java

```diff
@@ -144,19 +145,44 @@ Object extractFromSource(Map<String, Object> map) {
         Object value = map;
         boolean first = true;
         // each node is a key inside the map
-        for (String node : path) {
+        for (int i = 0; i < path.length; i++) {
+            String node = path[i];
             if (value == null) {
                 return null;
             } else if (first || value instanceof Map) {
                 first = false;
-                value = ((Map<String, Object>) value).get(node);
+                map = ((Map<String, Object>) value);
+                value = map.get(node);
+                if (value == null) { // Try to extract field with dots (e.g.: "b.c")
+                    Tuple<Object, Integer> tuple = extractAsDottedField(map, i, node);
+                    value = tuple.v1();
+                    if (value != null) {
+                        if (value instanceof Map) {
+                            i = tuple.v2();
+                        } else {
+                            return unwrapMultiValue(value);
+                        }
+                    }
+                }
             } else {
                 throw new SqlIllegalArgumentException("Cannot extract value [{}] from source", fieldName);
             }
         }
         return unwrapMultiValue(value);
     }
 
+    private Tuple<Object, Integer> extractAsDottedField(Map<String, Object> map, int idx, String node) {
```

Comment:
I'm thinking to remove this extracted code back to the body of `extractFromSource()` as now we create an extra `Tuple` and `Integer` object which happens for every document returned. What do you think?

Also the code might be a bit ugly but I wanted to avoid calling recursively a method (with every combination of `a.b`, `a.b.c`, etc.) Happy to hear better suggestions.

Reply:
+1. It looks like a small method and while it might be inlined the extra `Tuple/Integer` are boiler-plate.
If the method gets unrolled, both the index and the found value will be available without wrapping/boxing.